packages/nimble/R/nimbleFunction_Rderivs.R

```diff
@@ -0,0 +1,422 @@
+  flatteningInfo <- list()
+  thisIndex <- 1
+  for(i in seq_along(wrt)){
+    arg <- nf[[wrtArgIndices[i] + 1]]
+    indText <- ''
+    if(length(wrtNames[[i]]) > 1){
+      indText <-  paste0('[', wrtNames[[i]][[2]])
+    }
+    argSym <- parse(text = paste0(deparse(arg), indText))[[1]]
+    dimInfo <- dim(eval(argSym, envir = fxnEnv))
+    if(is.null(dimInfo)) dimInfo <- length(eval(argSym, envir = fxnEnv))
+    lineInds <- thisIndex:(thisIndex + prod(dimInfo) - 1)
+    argDimInfo <- dim(eval(nf[[wrtArgIndices[i] + 1]],envir = fxnEnv))
+    flatteningInfo[[i]] <- list()
+    flatteningInfo[[i]][[1]] <- argDimInfo
+    flatteningInfo[[i]][[2]] <- indText
+    flatteningInfo[[i]][[3]] <- lineInds
+    thisIndex <- thisIndex + prod(dimInfo)
+  }
+  wrappedFun <- function(x) {
+    args <- list()
+    for(i in 1:(length(nf)-1)){
+      if(i %in% wrtArgIndices){
+        thisWrtIndex <- which(i == wrtArgIndices)
+        thisSize <- sum(sapply(flatteningInfo[thisWrtIndex], function(x){return(prod(x[[1]]))})) ## total length
+        if(length(wrtNames[[thisWrtIndex[1]]]) > 1){
+            args[[i]] <-  eval(nf[[i+1]], envir = fxnEnv)
+            argBracketExpr <- parse(text = paste0('args[[i]]', flatteningInfo[[thisWrtIndex[1]]][[2]], ' <- x[flatteningInfo[[thisWrtIndex[1]]][[3]]]'))[[1]]
+            eval(argBracketExpr)
+            if(length(thisWrtIndex) > 1){
+              for(j in 2:length(wrtNames[thisWrtIndex])){
+                argBracketExpr <- parse(text = paste0('args[[i]]', flatteningInfo[[thisWrtIndex[j]]][[2]], ' <- x[flatteningInfo[[thisWrtIndex[j]]][[3]]]'))[[1]]
+                eval(argBracketExpr)
+              }
+            }
+          }
+          else{
+            args[[i]] <-  x[flatteningInfo[[thisWrtIndex[1]]][[3]]]
+          }
+          if(length( flatteningInfo[[thisWrtIndex[1]]][[1]]) > 1 ) dim(  args[[i]]) <-  flatteningInfo[[thisWrtIndex[1]]][[1]]
+      }
+      else{
+       args[[i]] <- nf[[i+1]] 
+      }
+    }
+    c(do.call(paste(nf[[1]]), args, envir = fxnEnv))
+  }
+  makeSingleArg <- function(){
+    singleArg <- c()
+    for(i in seq_along(wrtNames)){
+      if(length(wrtNames[[i]]) > 1){
+        arg <- eval(nf[[wrtArgIndices[i]+1]], envir = fxnEnv)
+        singleArg <- c(singleArg, c(eval(parse(text =  paste0('arg', flatteningInfo[[i]][[2]]))[1])))
+      }
+      else{
+        singleArg <- c(singleArg, c(eval(nf[[wrtArgIndices[i]+1]], envir = fxnEnv)))
+      }
+    }
+    return(singleArg)
+  }
+  return(list(wrappedFun, makeSingleArg))
+}
+
+
+#' Nimble Derivatives
+#' 
+#' EXPERIMENTAL Computes the value, gradient, and Hessian of a given  \code{nimbleFunction} method.  
+#' 
+#' @param nimFxn a call to a \code{nimbleFunction} method with arguments included.
+#' @param order an integer vector with values within the set {0, 1, 2}, corresponding to whether the function value, gradient, and Hessian should be returned respectively.
+#' 
+#' @export
```

Comment:
Add `@return`

Reply:
Just added some more documentation for this function, thanks @fritzo!